pkg/cmd/server/origin/run_components.go

```diff
@@ -376,7 +376,7 @@ func (c *MasterConfig) RunDeploymentImageChangeTriggerController() {
 // RunSDNController runs openshift-sdn if the said network plugin is provided
 func (c *MasterConfig) RunSDNController() {
 	oClient, kClient := c.SDNControllerClients()
-	controller, err := sdnfactory.NewMasterPlugin(c.Options.NetworkConfig.NetworkPluginName, oClient, kClient)
+	controller, err := sdnfactory.NewMasterPlugin(c.Options.NetworkConfig.NetworkPluginName, oClient, kClient, c.EtcdHelper)
```

Comment:
On Jun 21, 2016, at 7:38 PM, Ravi Sankar Penta <notifications@github.com>
wrote:

In pkg/cmd/server/origin/run_components.go
<https://github.com/openshift/origin/pull/8827#discussion_r67972369>:

> @@ -376,7 +376,7 @@ func (c *MasterConfig) RunDeploymentImageChangeTriggerController() {
>  // RunSDNController runs openshift-sdn if the said network plugin is provided
>  func (c *MasterConfig) RunSDNController() {
>  	oClient, kClient := c.SDNControllerClients()
> -	controller, err := sdnfactory.NewMasterPlugin(c.Options.NetworkConfig.NetworkPluginName, oClient, kClient)
> +	controller, err := sdnfactory.NewMasterPlugin(c.Options.NetworkConfig.NetworkPluginName, oClient, kClient, c.EtcdHelper)

I see these issues with the suggested approach:

   - No change to clients: This is undesirable (1) Old client calls are
   synchronous and to ensure backward compatibility, server needs to poll to
   get assigned vnid for namespace (we want this to happen on the client side).


So current clients don't retry if there is no vnid?  That's broken and we
should fix that before 3.3 by releasing a fix in 3.2.1.  That's a much
smaller risk migration issue.



   - (2) This introduces unnecessary indirection and complicates feature
   addition. Upcoming features like isolate-projects (undo join-projects) and
   ingress policy(upstream) stores info as annotations on namespace. To
   support this functionality via pod-network cli cmd in this case will need
   to pass annotation to NetNamespace first and then propagate to Namespace.


No, you would just call the net namespace API, and the registry would
handle it.  We've dealt with scenarios like this in the past and this is
the most manageable approach we've found.



   - I guess migrator is expected to run after upgrade of all nodes and
   masters. Once migrator is called, it will perform copy to namespace and
   deletes object but any create calls after that will still create etcd based
   NetNamespace objects? How long we will maintain this?


Until the admin sets a config flag, or we switch the controller to allocate
only via namespace.  Who calls create today?




NetNamespace is very specific to multitenant sdn plugin (unfortunately we
didn't enable this resource conditionally like builds). This was exposed to
support vnid watches on nodes and controlled updates by pod-network CLI. We
don't expect/want admin users to explicitly create/delete this resource, if
they do then there could be problems with isolation.


This is about mitigating risk during setup by avoiding both downtime and
the need for direct writes to etcd.

If we don't want controller to have direct access to etcd then we could do:

   - By default, make NetNamespace act as virutal resource backed by
   namespace. Old client works without any issues.


Only after a successful copy of the data from net namespaces to namespaces,
which would be incomplete.  Old client only works if the data is in place,
so this assumes a pre upgrade copy for existing namespaces.



   - Change List/Update/Delete methods to handle custom flag and when this
   flag is enabled, it should list/delete etcd backed netnamespaces. This will
   be used for migration.
   - New client hits Namespaces, there is no dependency on obsoleted
   NetNamespace.


That's fine.  New client has to wait if namespace doesn't have the
annotation though.


   - New nodes watches for both NetNamespaces and Namespaces, so new nodes
   can work with old masters.

Nodes can't really watch namespaces - this is a scalability problem (nodes
should not be holding in memory all namespaces).  You have to fetch
namespaces on demand and cache if necessary.  I'd actually prefer from a
security perspective if nodes didn't have access to namespaces, because
there is no guarantee namespace info isn't sensitive.  Nodes need to be
prevented from seeing that info - that's why NetNamespaces should continue
to exist as a virtual resource even after we've migrated away the etcd data.


   - New master will migrate NetNamespaces (client using custom flag)
   [Migration order: nodes and then masters] This will be much cleaner for
   future changes. No down time during migration, maintains backward
   compatibility and don't need to deal with etcd backed NetNamespaces after
   migration (no dual sets of values from namespaces and netnamespaces).


How is the master going to migrate that without access to etcd?  It would
be better if update on the NetNamespaces ensured that the data was written
to the namespace, or if we used that flag on the masters.

Ultimately, all schema migrations are four phases:

1. Write both old and new values (in this case etcd object and onto
namespace)
2. When all code is upgraded, finish upgrading the old data (migrator via
the API)
3. Flip a switch to only read and write new data (via config or etcd).
4. Delete old data

Nodes shouldn't have to change in this migration (except maybe to retry as
noted).  Keep the client API consistent.

Use the internal flag or etcd value to alter the behavior of the registry
to make the API alter.  Write both paths (forcing client to retry).  Run
migrator once all apiservers are upgraded.  Flip config flag once
everything is migrated.  Once all new API servers are running with the new
flag, delete old data from etcd.




For (b), Agree VNIDAllocationController/SecurityAllocationController
doesn't need etcd backing, I will change that to use in memory allocation.

@knobunc <https://github.com/knobunc> @dcbw <https://github.com/dcbw>
@danwinship <https://github.com/danwinship> @rajatchopra
<https://github.com/rajatchopra>
@liggitt <https://github.com/liggitt> @smarterclayton
<https://github.com/smarterclayton> any objections?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/8827/files/19fadc5fa9908cba956119f30049f34a67efecf6#r67972369>,
or mute the thread
<https://github.com/notifications/unsubscribe/ABG_p1Dg-outg-ZjvKwOHsVXC5dprHPoks5qOHXwgaJpZM4IblKD>
.